---
title: "Lab 4 - Analyzing Proportions"
output:
 html_document:
    theme: readable
    highlight: haddock
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir =  "~/Dropbox/Biometrics Labs/Lab 4 - Analyzing Proportions")
```

___

## Lab Objectives:
- Use R to conduct binomial testing
- Use R to compute a complete binomial distribution
- Use R to calculate confidence intervals for proportions using the Agresti-Coull method
- Practice using R

___


# Exercise 1: Binomial testing in R

### The binomial distribution

The binomial distribution describes the probability of a given number of successes out of n trials, where each trial independently has a p probability of success. It is given by the following equation:

###           $$Pr\left(X \right) = \binom{n}{X}p^{X}\left( 1-p \right)^{n-X}$$

X = total number of _successes_ (pass or fail, heads or tails etc.)

p = probability of a success on an individual trial

n = number of trials

The binomial test compares the observed number of successes in a data set to that expected under a null hypothesis. Today we will learn how to conduct this test in R.

<br>

___


### Remember to check or set your working directory 

Before getting started, set your working directory to where the course files are located.  The code below is commented out, since you will have to provide your own working directory.

```{r, echo=FALSE}
setwd("~/Dropbox/Biometrics Labs/Lab 4 - Analyzing Proportions")
```

```{r}
# setwd("Path to your files") # i.e. C:/Users/YourName/Documents/Biometrics Labs/
# note: this depends on your computer's OS
```

In RStudio, you can also change your working directory using the graphical interface:

![Set your working directory in RStudio](WorkingDirectory.jpg)

___
<br>

# Ear turn

Do people typically use a particular ear preferentially when listening to strangers? To answer this question, a researcher approached and spoke to strangers in a noisy nightclub, and an observer recorded whether the person turned their left or right ear toward the researcher. This data is found in the file Earturn.csv.


Then, load the file Earturn.csv into a variable called 'd', and examine it's data structure using the __structure()__ function and the __head()__ or __tail()__ functions to get a quick glance at the data.  Head shows the top 6 rows, tails displays the bottom 6 rows:

```{r}
d<-read.csv("Earturn.csv")
str(d)
head(d)
tail(d)
```

You should see that d is loaded as a data.frame, which is R terminology for a means of storing data in a table like format.  In this case, the data appear to be 25 observations with only 1 variable.  In other words, there are 25 rows and 1 column of information.

What the output also notes is that the data.frame contains 1 variable that called "Ear", which is listed a Factor w/ 2 levels: "left", "right".  This means, then that the file contains row-based entries on each measurement, whether it is right or left.  This is a long form way of encoding data and will need to be summarised before the binomial tests can be applied.

Perform an Exact Binomial Test, using the __binom.test()__ function. This function performs an exact test of a simple null hypothesis about the probability of success in a Bernoulli experiment.

However, the function requires that it be provided with a "vector of length 2, giving the numbers of successes and failures". 

We will use the __xtabs()__ function to perform a cross tabulation of the data and count up how many right and left turns there are. The tilde (~) is often used to specify groups or special formulae in R.  Putting the results of the __xtabs()__ function into the new variable success.fail helps to follow the example:

```{r}
xtabs(~d$Ear)
success.fail<-xtabs(~d$Ear)
success.fail
binom.test(success.fail, p = 0.5, conf.level = 0.95)
```

There are often multiple ways to do things in R.  You can manually specify x and n and input these as arguments in the function, binom.test:

```{r}
X<-6
n<-19+6
binom.test(X, n, p = 0.5, conf.level = 0.95)
``` 
 
If you want to simply have the p value returned, create a result variable and then call p value from the result, but using the dollar ($) sign in R, which is used typically to specify a named element (p.value) that is stored in your variable of interest.  

```{r}
result<-binom.test(X, n, p = 0.5, conf.level = 0.95)
result$p.value
```

Return to the data and change one right ear to a left ear. Re-run the Binomial test. Is the difference still significant? If so, return to the data and change another right ear to a left ear and re-run the binomial test until you find the point at which the proportion is no longer significantly different from the null hypothesis.

Note: the that Whitlock and Schluter advise that the p value returned from R's binom.test is not the same as one determined empirically, and often the solution is to multiply the number by 2. (see. p 196 Whitlock and Schluter re: p values and binomial tests.)

```{r}
success.fail<-c(7,18) # 6+1 and 19-1
result<-binom.test(success.fail, p = 0.5, conf.level = 0.95)
result$p.value*2
``` 
 
Note that you can also change the test proportion according to the null hypothesis. The null hypothesis for the ear turned is that there is no difference between left and right ears, so 0.5 is appropriate. However, this may change for other tests.

<br> 

#### Answer question 1 on Sakai

```{r, echo=FALSE}
success.fail<-c(8,17)
result<-binom.test(success.fail, p = 0.5, alternative = c("less"), conf.level = 0.95)
answer1a<-17/(8+17)
answer1b<-result$p.value*2 
# (see. p 196 Whitlock and Schluter re: p values and binomial test confusion...)
``` 
 
<br>

___

# Spermatogenesis genes

A study of genes involved in spermatogenesis was carried out to test the hypothesis that spermatogenesis genes should occur disproportionately often on the X chromosome, which contains 6.1% of the total genes. The researchers found that 40% of the analyzed spermatogenesis genes were on the X chromosome. This data is found in the file Spermatogenesis.csv.  

Read in the data, summarise the number of genes per chromosome using the __xtabs()__ function or the __table()__ function and plot the results using __barplot()__.

```{r}
d<-read.csv("Spermatogenesis.csv")
counts<-xtabs(~d$chromosome)
barplot(counts, beside=T, xlab="Chromosome", ylab="Number of genes")

counts<-table(d$onX,d$chromosome )
counts

barplot(counts, beside=T, xlab="Chromosome", ylab="Number of genes", 
        legend=rownames(counts), ylim=c(0,14))

```

The barplot function can handle data that are grouped.  In this case, by the onX variable, which yields 'no' or 'yes' as potential categories.  Note: this is a strange way to plot your data, as the grouping variable is redundant.

Now, peform a binomial test of the frequency of genes on the X chromosome relative to the null expectation:

```{r}
X=10  
n=sum(counts)
p=0.061
p.value<-binom.test(X, n, p)$p.value
p.value             # one-tailed p value      
p.value*2           # two-tailed equivalent

```


Now imagine we are interested in the proportion of spermatogenesis genes on the Y chromosome.  The Y chromosome contains only 1.2% of the total genes in the mouse genome. Make use of the __sum()__ command and a boolean expression using the __==__ command to count the observed number of genes on the Y, and compare to that expected under the null hypothesis. 

```{r}
d$chromosome=="Y"           # this should return a series of TRUE and FALSE values
sum(d$chromosome=="Y")      # sum up the number of TRUE statements
X=sum(d$chromosome=="Y")
n=sum(counts)
# Now run the appropriate binomial test to answer the question
```

Make note of the observed proportion of genes on the Y chromosome and the P-value for a two-tailed test.

<br> 

#### Answer question 2 on Sakai

```{r, echo=FALSE}
# Answer to question 2
answer2a<-X/n
p=0.012
p.value<-binom.test(X, n, p)$p.value
answer2b<-binom.test(X, n, p, alternative="greater")$p.value*2
```

<br>

___


# Groundhog Prognostication

![Groundhog Day](GHKissTop.jpg)

According to <http://www.stormfax.com/ghogday.htm>, the groundhog's ability to predict spring is about 39% accurate, even though the actual spring data were not provided in the link.

```{r}
d<-read.csv("GroundhogDay.csv")
str(d)
counts<-table(d$Result)
sum(counts)
```

From the table above, there are 10 records with no data and 130 records in total 
thus n = 130-10 = 120, and the success rate is 39% or 0.39, as reported on the website above.

```{r}
n=130-10
X=0.39*n
```

Note this yields a fraction value for X, not a whole number, so we are not working with the actual data, but the binomial test below will return an error if we do not provide whole numbers, so we will have to round the x off to the nearest integer before running the binomial test, to test whether the groundhog's predictions differ from a 50% random chance prediction:

```{r}
X<-round(0.39*n)
X
binom.test(X, n, p=0.5, alternative="two.sided")$p.value
```

Punxsutawney Phil's prediction rate falls significantly below that of random chance!! 

However, see a more detailed mathematical and scientific analysis here that suggests the groundhog has become a much more savvy groundhog, acheiving ~70% accuracy between 1955-1999:<br>

<https://www.jstor.org/stable/pdf/2687216.pdf?refreqid=excelsior%3Af9de80994249fe0e5103a0399b8b1e38>

<br> 

___

# Lost Wallets

In a study in Scotland, researchers left a total of 240 wallets around Edinburgh, as though the wallets were lost. Each contained contact information including an address. Data on the number of wallets that were returned is found in the file Wallets.csv.

Load the file in using code derived from the examples above.

<br>

#### Answer question 3 and 4 on Sakai

```{r, echo=FALSE}
d<-read.csv("Wallets.csv")
# str(d)
answer3a<-sum(d$wallet_returned=="returned") # 101
answer3b<-answer3a/sum(counts)  # 0.42
counts<-table(d)

p=0.5
answer3c<-binom.test(counts, p, alternative = "two.sided")$p.value # 0.0167
result<-binom.test(counts, p, alternative = "greater")$p.value*2

answer4<-answer3c<0.05 # TRUE
```

<br> 

___

# Exercise 2: Computing a binomial distribution in R

## Choosing wines

One function of R that we have not yet examined is its use to compute various statistical values as new variables. We previously examined an example in which 15 participants chose between two types of wine (cheap and expensive). If there was no preference between the two types of wine, then we would expect the proportion of individuals preferring each type of wine to be approximately 0.5 (50%).

The binomial distribution provides the probability distribution for the number of successes in a fixed number of trials, when the probability of success is the same in each trial. Applied to our example, the binomial distribution can tell us the probability of having a particular number of participants (e.g. 4, 8, 14) prefer the expensive wine over the cheap wine.

We will use the __dbinom()__ function to return a result of a binomial distribution, wherein we can obtain the probability of obtaining a particular number of 'successes' (i.e., 4, 8, or 14). 


```{r}
n=15
p=0.5
```

Probability of 4 people

```{r}
NumPick<-4
dbinom(NumPick, size=n, prob=p)
```

Probability of 8 people

```{r}
NumPick<-8
dbinom(NumPick, size=n, prob=p)
```

Probability of 14 people

```{r}
NumPick<-14
dbinom(NumPick, size=n, prob=p)
```

Or, if you were interested in a range of probabilities, i.e. number >= 4:

```{r, echo=F}
ranbinom<-rbinom(100000, 15, 0.5)
h <- hist(ranbinom,  plot=FALSE)
cuts <- cut(h$breaks, c(-Inf,3,Inf))
hist(ranbinom, probability=T, col=cuts, main="Probability of >= 4 people in red",
     xlab="Number of people who prefer expensive wine")
```

You would need to sum across the densities obtain for each of those values.  In this case, summing the density from 4 to 15 will encompass all possibilities, since you cannot have more than 15 out of 15:

```{r}
probs<-dbinom(4:15, size=n, prob=p)
sum(probs)
```

<br> 

#### Answer question 5 on sakai 

```{r, echo=FALSE}
answer5<-binom.test(x=10, n=15, p=0.5)
```
<br>

___

## Heterozygous crosses

In a cross between two heterozygotes, 25% of the offspring is expected to be homozygous recessive. In pea plants, green pod colour is dominant, and yellow pod colour is recessive. In a cross between two heterozygous pea plants you obtain 20 offspring.

Compute the binomial distribution for obtaining 10 yellow plants out of n = 20.

```{r}
dbinom(x=10, size=20, p=0.25)
```

<br>

#### Answer question 6 on sakai

```{r, echo=FALSE}
answer6<-dbinom(x=20/4, size=20, p=0.25)
```

<br>

___

# Exercise 3: 95% Confidence intervals for proportions

To practice the Agresti-Coull method for calculating the 95% confidence interval of a proportion, we will need the following equations:

### $$p' = \frac{X+2}{n+4}$$

<br>

### $$p' - 1.96\sqrt{\frac{p'\left(1-p'\right)}{n+4}} < p < p' + 1.96\sqrt{\frac{p'\left(1-p'\right)}{n+4}}$$

<br> 

___

## Murphy's law

In a test of Murphy‘s Law, pieces of toast were buttered on one side and then dropped. Murphy’s Law predicts that they will land butter-side down. Out of 9821 slices of toast dropped, 6101 landed butter-side down.

Open the file Butter.csv, which contains the data from the Murphy ’s Law experiment described above.

```{r}
d<-read.csv("Butter.csv")
str(d)
levels(d$sideDown)
sum(d$sideDown=="butter side")
X<-6101
n<-9821
pp<-(X+2)/(n+4)
Conf.Range<-1.96*sqrt(pp*(1-pp)/(n+4))

```

Compute the confidence intervals and perform a binomial test on the data.

<br>

#### Answer questions 7 and 9 on sakai

```{r, echo=FALSE}
answer7a<-X/n # proportion butter side down 0.6212
lower.limit<-X/n - Conf.Range
upper.limit<-X/n + Conf.Range
answer7b<-c(lower.limit, upper.limit)
answer9<-binom.test(X, n, p=0.5)$p.value<0.05
```

<br>

___

## Surgical infections

Out of 67410 surgeries tracked in a study in the UK, 2832 were followed by surgical site infections.

Calculate the 95% CI for the proportion of surgical site infections. 

If a 95% confidence interval does not include the value provided in the null hypothesis, then the null hypothesis will usually be rejected. This makes sense, as it indicates that the value of the null hypothesis is not one of the most plausible values given the data.

```{r}
X<-2832
n<-67410
pp<-(X+2)/(n+4)
Conf.Range<-1.96*sqrt(pp*(1-pp)/(n+4))
```

<br>

#### Answer question 8 on sakai

```{r, echo=FALSE}
answer8a<-X/n  # 0.042 = proportion of surgical infections
lower.limit<-X/n - Conf.Range
upper.limit<-X/n + Conf.Range
answer8b<-c(lower.limit, upper.limit) # 0.0405 to 0.0435
```

<br>

___

# Exercise 4: R Practice 

## Country Data

A variety of population statistics collected from 209 different countries is found in the file Countries.sav.

Open the file Countries.csv

```{r}
d<-read.csv("Countries.csv")
str(d)
anyNA(d) 
sum(is.na(d))
```

Unfortunately there are quite a few missing data points in the file, as the function __anyNA()__ and __is.na()__ reveal.  This may present some challenges.  You may need to include na.rm=T when calling mean, sum, or other functions.

```{r}
mean(d$Agricultural_land)
mean(d$Agricultural_land, na.rm=T)
```

DPT is a combination vaccine against diphtheria, pertussis and tetanus. Use the __hist()__ function to plot a histogram for the Immunization DPT variable, which contains data on the % of children ages 12 – 23 months who received this vaccine in different countries. 

```{r}
hist(d$Immunization_DPT, xlab="Percent children immunised with DPT")
```

Change the binning to a custom interval width of 10, by setting the breaks = seq(0,100,10):

```{r}
hist(d$Immunization_DPT, breaks=seq(0,100,10))
```

Histograms can also be returned as objects to work from.  Typically you would also set plot=FALSE when doing this:

```{r}
h<-hist(d$Immunization_DPT, plot=FALSE)
str(h)
h$breaks
h$counts
```

Use this function to generate histograms for personal computers per 100 people, and internet users per 100 people.

To count up particular information, use the sum and boolean commands:

```{r}
sum(d$Immunization_DPT>90, na.rm=T) # number of countries with >90% immunisation
sum(d$Immunization_DPT>=70 & d$Immunization_DPT<80, na.rm=T) # number of countries with immunisation rates >= 70 and <80%
```

If we are interested in the descriptive statistics (values) only, we may choose to use the function available in a different package, such as __stat.desc__ from the <pastecs> package. 

[You will likely have to install the pastecs package first (remove the # from in front of install.packages("pastecs").]

Use the __stat.desc__ function to produce a table containing mean, standard deviation, minimum, maximum, and SE mean for the three variables examining life expectancy at
birth (male, female, and total).

```{r}
# install.packages("pastecs")
library(pastecs)
options(scipen=9)
variable<-d$Immunization_DPT
cbind(stat.desc(variable, basic=T))
```

The variable population_female indicates the percent of the total population that is female. Calculate mean and 95% confidence interval for this variable.

<br> 

#### Answer questions 10-14 on Sakai

<br> 

```{r, echo=FALSE}
answer10<-mean(d$Immunization_DPT, na.rm=T) # mean % children immunised is 86.5
answer11<-"skewed left" 
# histogram for dpt immun is left skewed.  they will need to # look up what skew means
h<-hist(d$Internet_users, plot=F,  breaks=seq(0,100,10))
answer12a<-sum(d$Internet_users>=60 & d$Internet_users<70, na.rm=T)
# The number of countries with 60-70 internet users per 100 people is 10.
h<-hist(d$Internet_users, plot=F, breaks=seq(0,100,10))
answer12b<-sum(d$Personalcomputers>=0 & d$Personalcomputers<10, na.rm=T)

# The number of countries with 0-10 personal computers per 100 people is 110.
answer13a<-mean(d$Life_expectancy_birth_female, na.rm=T)
answer13b<-mean(d$Life_expectancy_birth_male, na.rm=T)
# The mean life expectancy at birth for females is 70.1. The mean life expectancy at birth for males is 65.4.
options(scipen=9)
results<-cbind(stat.desc(d$Population_female))

answer14a<-c(49.89-0.34648,49.89+0.34648)
# The 95% confidence interval for the % of the population that is female is 49.55 < μ < # 50.24.
answer14b<-32
# The minimum percent female observed in any country is 32%

```
___


